neps/nep-0455.md

```diff
@@ -0,0 +1,219 @@
+If processing the chunk takes much longer than 1.3s, it can cause the corresponding block and possibly more consecutive blocks to be skipped.
+
+We could extend the protocol to produce empty chunks for some of the shards within the block (effectively skipping them) when processing the chunk takes longer than expected.
+This way will still ensure a regular block production, at a cost of lower throughput of the network in that shard.
+The chunk should still be included in a later block to avoid stalling the affected shard.
+
+Pros:
+- Fast and automatic adaptation to the blockchain workload
+
+Cons:
+- Needs a voting mechanism to decide that processing of the chunk is slow (as opposed to the chunk producer being slow) and the chunk should be skipped within a block
+- Needs economic design for the rewards for skipped chunks
+
+## Specification
+
+- **Chunk Compute Usage** -- total compute time spent on processing the chunk
+
+- **Chunk Compute Limit** -- upper-bound for compute time spent on processing the chunk
+
+- **Parameter Compute Cost** -- the numeric value in seconds corresponding to compute time that it takes to include an operation into the chunk
+
+Today, gas has two somewhat orthogonal roles:
+
+1. Gas is money. It is used to avoid spam by charging users
+2. Gas is CPU time. It defines how many transactions fit in a chunk so that validators can apply it within a second
+
+The idea is to decouple these two by introducing parameter compute costs.
+Each gas parameter still has a gas cost that determines what users have to pay.
+But when filling a chunk with transactions, parameter compute cost is used to estimate CPU time.
+
+Ideally, all compute costs should match corresponding gas costs.
+But when we discover undercharging issues, we can set a higher compute cost (this would require a protocol upgrade).
+The stability concern is then resolved when the compute cost becomes active.
+
+The ratio between compute cost and gas cost can be thought of as an undercharging factor.
+If a gas cost is 2 times too low to guarantee stability, compute cost will be twice the gas cost.
+A chunk will be full 2 times faster when gas for this parameter is burned.
+This deterministically throttles the throughput to match what validators can actually handle.
+
+Compute costs influence the gas price adjustment logic described in https://nomicon.io/Economics/Economic#transaction-fees.
+Specifically, we're now using compute usage instead of gas usage in the formula to make sure that the gas price increases if chunk processing time is close to the limit.
+
+Compute costs **do not** count towards the transaction/receipt gas limit of 300TGas, as that might break existing contracts by pushing their method calls over this limit.
+
+Compute costs are static for each protocol version.
+
+### Using Compute Costs
+Compute costs different from gas costs are only a temporary solution.
```

Comment:
I appreciate the intention of having different gas cost from compute cost to be temporary, but we should assume that the temporal period can be large, or even at some points permanent.

I recommend suggesting in this PR a workflow for developers that plan to change the compute cost, about how they plan it to be fixed (i.e for making gas cost match compute cost again). Maybe opening an issue in NEPs repository is ok, or adding entries on a new section of this NEP. So we can keep track off all current differences, and see what is the remediation plan.

What are the implications of having this difference permanently? 
- Metrics will be affected. (i.e gas cost != compute cost so blocks will be seen emptier than they really are)
- We are currently exposing profiling information of tx execution through the RPC. Should we expose both compute cost and gas cost? (Not sure about this one).